### ISSUE TITLE
[dtype conversion BUG] Data type conversion may cause blocking.

### ISSUE THREAD

**Issue by arminzhu** [pytorch/pytorch#185810]
### 🐛 Describe the bug

I encountered a bug related to data type conversion. The `Tensor.to()` can cause blocking.

The code:

```
import torch
import time

def compute_ops(A, B):
    h=A.mm(A)
    m=B.mm(B)
    return h.mm(m)

if __name__ == '__main__':
    torch.cuda.set_sync_debug_mode("warn")
    _device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    tensor_A = torch.randn(16384,16384, device=_device,dtype=torch.float32,requires_grad=True)
    tensor_B = torch.randn(16384,16384, device=_device,dtype=torch.float32,requires_grad=True)
    tensor_C = torch.randn(8,8, device=_device,dtype=torch.float32).contiguous()

    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    
    start_time = time.time()
    start.record()
    out=compute_ops(tensor_A, tensor_B)
    end.record()

    tensor_16 =  tensor_C.to(dtype=torch.bfloat16,non_blocking=True)
    
    print(f"Python time: {time.time() - start_time}")
    torch.cuda.synchronize()
    print(f"Event elapsed_time: {start.elapsed_time(end)/1000.}")
```

The result:
```
Python time: 2.41646671295166
Event elapsed_time: 2.41256201171875
```

### Versions

Software versions:
    PyTorch 2.10
    CUDA 12.8

Hardware:
    r9 7945HX
    RTX 5060 Laptop

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

**Comment by arminzhu** [comment]
If we remove the line of code `tensor_16 = tensor_C.to(dtype=torch.bfloat16, non_blocking=True)`, then the `Python time` will be extremely short.

**Comment by dt1dr** [comment]
Follow-up with additional findings from reproducing this.

**Environment**
- GPU: Tesla T4 (Colab)
- CUDA: 12.8
- PyTorch: 2.10.0+cu128
- OS: Linux

A few observations:

The stall only shows up on the **first** invocation in a fresh process. Any subsequent dtype conversion runs cleanly.

So I tried toggling the module loading mode:

| Mode | Behavior |
|---|---|
| `CUDA_MODULE_LOADING=LAZY` (default) | Python time ≈ Event time (the stall) |
| `CUDA_MODULE_LOADING=EAGER` | 1s < Python time << Event time (expected) |

Under EAGER, `non_blocking=True` behaves as expected: Python returns right away while the GEMM keeps running in the background. 
I also tried adding a single warmup dtype conversion before the measurement under LAZY mode, and that makes the stall disappear too, which lines up with what the CUDA Programming Guide recommends.

Section 4.7.5.3 ("Impact on Performance Measurements") basically describes this exact scenario that lazy loading can pull module initialization into the first measured launch:
https://docs.nvidia.com/cuda/cuda-programming-guide/04-special-topics/lazy-loading.html

Given that it:
- only happens once per process,
- goes away under `CUDA_MODULE_LOADING=EAGER`, and
- goes away with a warmup,

This is likely a driver-level lazy loading rather than `non_blocking` actually misbehaving. Probably worth documenting as a known interaction, but I don't think it points to an issue on the PyTorch side.

Happy to run anything else if it'd be useful :)

**Comment by arminzhu** [comment]
@dt1dr Thank you very much for your answer, it helped me a lot. I will close the issue.
